File: ios/Classes/FlutterAria2Plugin.m
#import "FlutterAria2Plugin.h"

#import "FlutterAria2Native.h"

@interface FlutterAria2Plugin ()
@property(nonatomic, strong) FlutterMethodChannel* channel;
@property(nonatomic, strong) FlutterAria2Native* native;
@end

@implementation FlutterAria2Plugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel =
      [FlutterMethodChannel methodChannelWithName:@"flutter_aria2"
                                  binaryMessenger:[registrar messenger]];
  FlutterAria2Plugin* instance = [[FlutterAria2Plugin alloc] initWithChannel:channel];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)initWithChannel:(FlutterMethodChannel*)channel {
  self = [super init];
  if (self) {
    _channel = channel;
    _native = [[FlutterAria2Native alloc] init];
    __weak typeof(self) weakSelf = self;
    _native.onDownloadEvent = ^(NSInteger event, NSString* gid) {
      [weakSelf.channel invokeMethod:@"onDownloadEvent"
                           arguments:@{
                             @"event" : @(event),
                             @"gid" : gid ?: @"",
                           }];
    };
  }
  return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  [self.native invokeMethod:call.method
                  arguments:[call.arguments isKindOfClass:[NSDictionary class]]
                                ? (NSDictionary<NSString*, id>*)call.arguments
                                : nil
                 completion:^(id _Nullable value, NSError* _Nullable error) {
                   if (error == nil) {
                     result(value);
                     return;
                   }
                   NSString* code = error.userInfo[@"code"];
                   if ([code isKindOfClass:[NSString class]] &&
                       [code isEqualToString:@"NOT_IMPLEMENTED"]) {
                     result(FlutterMethodNotImplemented);
                     return;
                   }
                   result([FlutterError errorWithCode:code ?: @"NATIVE_ERROR"
                                              message:error.localizedDescription
                                              details:nil]);
                 }];
}

@end

File: macos/Classes/FlutterAria2Plugin.swift
import Cocoa
import FlutterMacOS

public class FlutterAria2Plugin: NSObject, FlutterPlugin {
  private let channel: FlutterMethodChannel
  private let native: FlutterAria2Native

  private init(channel: FlutterMethodChannel) {
    self.channel = channel
    self.native = FlutterAria2Native()
    super.init()

    native.onDownloadEvent = { [weak channel] event, gid in
      channel?.invokeMethod("onDownloadEvent", arguments: [
        "event": event,
        "gid": gid,
      ])
    }
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_aria2", binaryMessenger: registrar.messenger)
    let instance = FlutterAria2Plugin(channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    native.invokeMethod(call.method, arguments: call.arguments as? [String: Any]) { value, error in
      guard let error else {
        result(value)
        return
      }

      let nsError = error as NSError
      let code = (nsError.userInfo["code"] as? String) ?? "NATIVE_ERROR"
      if code == "NOT_IMPLEMENTED" {
        result(FlutterMethodNotImplemented)
        return
      }
      result(FlutterError(code: code, message: nsError.localizedDescription, details: nil))
    }
  }
}
